stratego/case-studies/Main.hs

```diff
@@ -23,137 +22,27 @@ import           Data.Foldable
---     k $ \grammar name fun depth res -> measure "Rule Invocations Analysis" $ forM_ res $ \t ->
---       case summary grammar t of
---         Just s -> forM_ (M.toList s `zip` [(1::Int)..]) $ \((rule,count),ruleId) ->
---           hPrintf csv "%s;%s;%d;%s;%d;%s;%d\n" name fun depth (show t) ruleId (show rule) count
---         Nothing -> return ()
-
--- type Distance = Int
-
--- | Classifies terms as true and false, positive or negatives.
---
--- True-Positive: Terms in the analysis result that are part of the output language of the program transformation
--- False-Positive: Terms in the analysis result that are *not* part of the output language of the program transformation
--- True-Negative: Terms that occur neither in the output language of the program transformation nor in the analysis result
--- False-Negative: Terms in output language of the program transformation that do not occur in the analysis result
--- classificationSetup :: ((Grammar -> Distance -> Analysis t) -> IO ()) -> IO ()
--- classificationSetup k =
---   withFile "classification.csv" WriteMode $ \csv -> do
---     hPrintf csv "name;fun;depth;sample_distance;term;height;size;distance_sum;class\n"
---     k $ \grammar sampleDistance name fun depth analysis -> measure "Classification" $ do
---       let relevant = termsOfDistance sampleDistance grammar
---           selected = H.delete W.Wildcard analysis
---           (truePositive,falsePositive) = partition (\s -> any (\r -> s `wittnesses` r) relevant) selected
---           (_,           falseNegative) = partition (\r -> any (\s -> s `wittnesses` r) selected) relevant
---       forM_ [("true_positive", truePositive), ("false_positive", falsePositive), ("false_negative", falseNegative)] $ \(klass,terms) ->
---         forM_ terms $ \t ->
---           hPrintf csv "%s;%s;%d;%d;%s;%d;%d;%d;%s\n"
---             name fun depth sampleDistance (show t) (T.height t) (T.size t) (distanceSum grammar t `orElse` 0) (klass :: String)
---   where
---     partition :: (Eq a, Hashable a) => (a -> Bool) -> HashSet a -> (HashSet a, HashSet a)
---     partition predicate = H.foldl' (\(pos,neg) t -> if predicate t then (H.insert t pos,neg) else (pos,H.insert t neg))
---                                    (H.empty,H.empty)
-
--- wittnesses :: W.Term -> W.Term -> Bool
--- wittnesses W.Wildcard W.Wildcard = True
--- wittnesses W.Wildcard (W.Cons _ []) = True
--- wittnesses W.Wildcard (W.StringLiteral _) = True
--- wittnesses W.Wildcard (W.NumberLiteral _) = True
--- wittnesses (W.Cons c ts) (W.Cons c' ts') = c == c' && length ts == length ts' && and (zipWith wittnesses ts ts')
--- wittnesses (W.StringLiteral s) (W.StringLiteral s') = s == s'
--- wittnesses (W.NumberLiteral n) (W.NumberLiteral n') = n == n'
--- wittnesses _ _ = False
-
--- for :: Functor f => f a -> (a -> b) -> f b
--- for = flip fmap
-
--- orElse :: Maybe a -> a -> a
--- orElse = flip fromMaybe
-
 measure :: String -> IO () -> IO ()
```

Comment:
Do you still want to measure this? Otherwise I should remove this function.

Reply:
The same goes for the line `_ <- CM.measure (CT.nfIO (return terms)) 1` in `caseStudy` below.

Reply:
It sets up a timer of how long the case studies run.

Reply:
It's unused at the moment.